### PR TITLE
Add Smart Contract guide

### PR DESCRIPTION
**Pull request by Arbaba**
Add tutorial on smart contract deployment.

### CODE REVIEW

File: Smart-Contract-Guide.md [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+
+## Create and deploy a script
+Now that the contract has been successfully created, we will deploy a `TxScript` which calls the `Mytoken.buy` method to obtain tokens by paying **ALPH** to the contract. For this example, we will pay using address `1ELgp7U4D1QL82G9q9dAdp43k45onPDezGLjHSFGcwCj9` which is different than the one used to create the contract.
+
+If you also want to pay with an address different than the one used to submit the contract, please make sure that your address belongs to the same group as the contract. You can obtain the contract group by checking the `chainFrom` field of it's transaction block. In our example, the contract is in group 0, but it might be in a different group for you. You can verify the group of an address at endpoint `GET addresses/{address}/group`. If it is not the case, you can use `POST/wallets/{wallet_name}/derive-next-address` until you obtain an address in the correct group. As new addresses are initialized with balance `0`, you should transfer some **ALPH** to this new address. Finally, change your active address at endpoint `POST wallets/{wallet_name}/change-active-address`.
```
Comment (polarker) [inline Smart-Contract-Guide.md]: of `its` transaction block

File: Smart-Contract-Guide.md [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+    approveAlf!(@1ELgp7U4D1QL82G9q9dAdp43k45onPDezGLjHSFGcwCj9, 1000000000000000000)
+    let contract = MyToken(#109b05391a240a0d21671720f62fe39138aaca562676053900b348a51e11ba25)
+    contract.buy(@1ELgp7U4D1QL82G9q9dAdp43k45onPDezGLjHSFGcwCj9, 1000000000000000000)
+  }
+}
+```
+Here is a brief explanation of this code:
+
+* `approveAlf!(address, amount)` authorizes the specified amount of `ALPH` from the address to be used in the script.
+* The contract is loaded by it's `id`, the previously obtained base58 encoded address
```
Comment (polarker) [inline Smart-Contract-Guide.md]: now the contract `id` is returned when we build the contract tx. No need to get it from base58

File: Smart-Contract-Guide.md [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+final case class ContractState private (
+    code: StatefulContract.HalfDecoded,
+    initialStateHash: Hash,
+    fields: AVector[Val],
+    contractOutputRef: ContractOutputRef
+)
+```
+where the fields are:
+
+
```
Comment (polarker) [inline Smart-Contract-Guide.md]: one empty is enough

**Review by polarker — APPROVED** [review]
Great work. Could you please add a sentence at the beginning of the end of the page to mention that it's based on @Lbqds 's Chinese doc ?

File: Smart-Contract-Guide.md [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+This document is based on the [Chinese smart contract tutorial and documentation](https://github.com/Lbqds/alephium-docs/blob/master/contract.md) by [Lbqds](https://github.com/Lbqds).
+
+## Create and deploy a token contract
+
+In this section we will create, build, sign and submit a contract transaction.
+
+### Create a token Contract
+First, we create a token contract as shown below:
+
+```rust
```
Comment (killerwhile) [inline Smart-Contract-Guide.md:25]: is this snippet really `rust` code?

File: Smart-Contract-Guide.md [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+    let tokenId = selfTokenId!()
+    transferAlf!(from, owner, alfAmount)
+    transferTokenFromSelf!(from, tokenId, tokenAmount)
+    remain = remain - tokenAmount
+  }
+}
+```
+This simple token contract allows users to buy tokens by paying **ALPH** to the contract owner at a rate `1:1000`. It uses the following built-in functions:
+
+* `assert!(pred)` Causes the contract exectuion to fail when `pred` evaluates to `false`
```
Comment (killerwhile) [inline Smart-Contract-Guide.md]: typo in exectuion -> execution

File: Smart-Contract-Guide.md [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+* `selfTokenId!()` Returns the current token id which is also the current contract id
+* `transferAlf(from, to, alfAmount)` Transfers `alfAmount` **ALPH** from address `from` to `to`.
```
Comment (killerwhile) [inline Smart-Contract-Guide.md]: `alfAmount` is a bit misleading, now with the new notation 1.2 ALPH.

File: Smart-Contract-Guide.md [hunk 7]
```diff
@@ -0,0 +1,606 @@
+# Smart Contract Guide
+
+This document guides you through the creation, deployment and usage of smart contracts on Alephium testnet.
+
+We will first deploy a contract which allows any user to exchange **ALPH** for tokens. Then we will deploy a script which calls the contract to buy tokens.
+
+This document is based on the [Chinese smart contract tutorial and documentation](https://github.com/Lbqds/alephium-docs/blob/master/contract.md) by [Lbqds](https://github.com/Lbqds).
+
```
Comment (killerwhile) [inline Smart-Contract-Guide.md:8]: I'd add a mention "Requirement to have a broker running locally", with a link to the doc of running a broker

File: Smart-Contract-Guide.md [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+  -H 'accept: application/json' \
+  -H 'Content-Type: application/json' \
+  -d '{
+  "code": "TxContract MyToken(owner: Address, mut remain: U256) {\n  pub payable fn buy(from: Address, alfAmount: U256) -> () {\n    let tokenAmount = alfAmount * 1000\n    assert!(remain >= tokenAmount)\n    let tokenId = selfTokenId!()\n    transferAlf!(from, owner, alfAmount)\n    transferTokenFromSelf!(from, tokenId, tokenAmount)\n    remain = remain - tokenAmount\n  }\n}"
+}'
+```
+
+We receive the binary code of the contract as a response:
+
+
```
Comment (killerwhile) [inline Smart-Contract-Guide.md]: can remove one new line

File: Smart-Contract-Guide.md (hunk 7 above)
Comment (killerwhile) [reply Smart-Contract-Guide.md:8]: with a wallet `demo-1`

File: Smart-Contract-Guide.md [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+```json
+{
+  "unsignedTx": "0101010101000000071500a273b4a26cef6181b90a80c5e8738478ed35b9ea10f9e00e12225174a86ebd8113c1e8d4a51000a21440300201402c01010204001616011343e82c1702a0011602344db117031600a0001601a7160016031602aba00116022ba101144031020400a273b4a26cef6181b90a80c5e8738478ed35b9ea10f9e00e12225174a86ebd8102c8204fce5e3e2502611000000013c8204fce5e3e25026110000000ae80013880c1174876e8000145827e150d3aa8643039f353c4e0f6bc554d53e61304f566f3302323c89eede0b86e130d00027b029462e7df54cd218cc46e2f3c8fd6cfd3c0d475474e5a3ded70d809df04a700",
+  "hash": "8d01198f2ec74b1e5cfd8c8a37d6542d16ee692df47700ce2293e0a22b6d4c22",
+  "contractId": "109b05391a240a0d21671720f62fe39138aaca562676053900b348a51e11ba25",
+  "fromGroup": 0,
+  "toGroup": 0
+}
+```
+### Sign a contract
```
Comment (killerwhile) [inline Smart-Contract-Guide.md:119]: Can add a new line here

File: Smart-Contract-Guide.md [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+
+This transaction has one input and two outputs. Below is the description of the some fields:
+* `outputRef` pointer to UTXO
+* `outputRef.key` UTXO hash
+* `type` the type of the tx output, `contract` or `asset`
+* `address` base58 encoding of the `contract` or `asset` address
+* `amount` the amount of ALPH owned by the address
+* `tokens` list of tokens owned by the address
+
+The first output is the contract we just created. We see that the contract address owns `10000000000000000000000000000` tokens which is exactly the `issueTokenAmount` we previously defined.
```
Comment (killerwhile) [inline Smart-Contract-Guide.md:309]: I'd add the path `tokens.amount`, since it's a bit misleading with `amount`

**Review by killerwhile — COMMENTED** [review]
Review part 1/2 :)

File: Smart-Contract-Guide.md (hunk 4 above)
Comment (polarker) [reply Smart-Contract-Guide.md:25]: nope, but it's similiar to `rust`, so @Arbaba used this hack

File: Smart-Contract-Guide.md [hunk 11]
```diff
@@ -0,0 +1,606 @@
...
+    approveAlf!(@1ELgp7U4D1QL82G9q9dAdp43k45onPDezGLjHSFGcwCj9, 1000000000000000000)
+    let contract = MyToken(#109b05391a240a0d21671720f62fe39138aaca562676053900b348a51e11ba25)
+    contract.buy(@1ELgp7U4D1QL82G9q9dAdp43k45onPDezGLjHSFGcwCj9, 1000000000000000000)
+  }
+}
+```
+Here is a brief explanation of this code:
+
+* `approveAlf!(address, amount)` authorizes the specified amount of `ALPH` from the address to be used in the script.
+* The contract is loaded by it's `id`
```
Comment (killerwhile) [inline Smart-Contract-Guide.md]: typo: it's -> its

**Review by killerwhile — APPROVED** [review]
Very few more comment. Great write up in general!

File: Smart-Contract-Guide.md [hunk 12]
```diff
@@ -0,0 +1,606 @@
...
+"issueTokenAmount": "10000000000000000000000000000"
+}'
+```
+
+The parameters are:
+
+* `fromPublicKey`  Public key from the address currently in use
+* `code` Contract binary code
+* `gas` Manually specified gas as the default gas may not be enough for contract related operations
+* `issueTokenAmount` The total number of tokens issued by the contract
```
Comment (tdroxler) [inline Smart-Contract-Guide.md:105]: would be nice to explain the `state` field too

**Review by tdroxler — APPROVED** [review]
only one minor comment otherwise LGTM, great job :+1:

**Comment by Arbaba** [comment]
Thank you all for the feedback :)

**Pull request merged**